### PR TITLE
Ubiquity connector

### PR DESCRIPTION
**Pull request by zapaz**
Ubiquity connector allowing deposits to get Ubiquity Bonding Shares 
```

    /**
     * @dev Deposit into Ubiquity protocol
     * @notice 3POOL (DAI / USDC / USDT) => METAPOOL (3CRV / uAD) => uAD3CRV-f => Ubiquity BondingShare
     * @notice STEP 1 : 3POOL (DAI / USDC / USDT) => 3CRV
     * @notice STEP 2 : METAPOOL(3CRV / UAD) => uAD3CRV-f
     * @notice STEP 3 : uAD3CRV-f => Ubiquity BondingShare
     * @param token Token deposited : DAI, USDC, USDT, 3CRV, uAD or uAD3CRV-f
     * @param amount Amount of tokens to deposit (For max: `uint256(-1)`)
     * @param durationWeeks Duration in weeks tokens will be locked (4-208)
     * @param getId ID to retrieve amt.
     * @param setId ID stores the amount of tokens deposited.
     */
```

Run test via ` npx hardhat test test/ubiquity/ubiquity.test.js`

First "DSA wallet setup" test setup and deposit with different types of tokens
Then "Main" test deposit from DSA to Ubiquity protocol to get Bonding Shares

**`Ubiquity Docs`**  https://dao.ubq.fi/docs

- [x] Review and approve connector @thrilok209 
- [x] Deploy the contracts and comment the addresses @thrilok209 
- [x] Created resolver - [here](https://github.com/Instadapp/dsa-resolvers/pull/15)

### CODE REVIEW

**Comment by thrilok209** [comment]
@zapaz Can you attach some docs related to integration?

**Comment by thrilok209** [comment]
@zapaz After depositing, how will the user withdraw?

**Comment by zapaz** [comment]
yes of course @thrilok209 , a withdraw function is needed, I will add with integration docs also

**Comment by zapaz** [comment]
> @zapaz After depositing, how will the user withdraw?

Withdraw functionnality added

**Comment by marcomow** [comment]
hi @thrilok209 just checking out is the pr looking good? Otherwise our snowball is stuck

File: contracts/mainnet/connectors/ubiquity/main.sol [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+				amount1,
+				tok[0] ? 0 : (tok[1] ? 1 : 2),
+				0
+			);
+		}
+
+		uint256 amount = getTokenBal(TokenInterface(token));
+
+		setUint(setId, amount);
+		_eventName = "Withdraw(address,uint256,uint256,uint256,address,uint256,uint256,uint256)";
```
Comment (thrilok209) [inline contracts/mainnet/connectors/ubiquity/main.sol]:
```suggestion
		_eventName = "LogWithdraw(address,uint256,uint256,uint256,address,uint256,uint256,uint256)";
```

File: contracts/mainnet/connectors/ubiquity/main.sol [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+		address bonding = ubiquityManager.bondingContractAddress();
+		approve(TokenInterface(UAD3CRVf), bonding, lpAmount);
+		uint256 bondingShareId = IUbiquityBondingV2(bonding).deposit(
+			lpAmount,
+			durationWeeks
+		);
+
+		setUint(setId, bondingShareId);
+
+		_eventName = "Deposit(address,address,uint256,uint256,uint256,uint256,uint256,uint256)";
```
Comment (thrilok209) [inline contracts/mainnet/connectors/ubiquity/main.sol]:
```suggestion
		_eventName = "LogDeposit(address,address,uint256,uint256,uint256,uint256,uint256,uint256)";
```

File: contracts/mainnet/connectors/ubiquity/events.sol [hunk 3]
```diff
@@ -0,0 +1,25 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.7.6;
+
+contract Events {
+    event Deposit(
```
Comment (thrilok209) [inline contracts/mainnet/connectors/ubiquity/events.sol]:
```suggestion
    event LogDeposit(
```

File: contracts/mainnet/connectors/ubiquity/events.sol [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+        address indexed userAddress,
+        address indexed token,
+        uint256 amount,
+        uint256 indexed bondingShareId,
+        uint256 lpAmount,
+        uint256 durationWeeks,
+        uint256 getId,
+        uint256 setId
+    );
+    event Withdraw(
```
Comment (thrilok209) [inline contracts/mainnet/connectors/ubiquity/events.sol]:
```suggestion
    event LogWithdraw(
```

**Comment by thrilok209** [comment]
Hey @zapaz @marcomow! Connector looks good! Added minor suggestions. 

Next step would be deploying the connector and posting the deployed connector address here. So, we will review it and whitelist the connector.

File: contracts/mainnet/connectors/ubiquity/main.sol [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+ * @title Ubiquity.
+ * @dev Ubiquity Dollar (uAD).
+ */
+
+import { TokenInterface } from "../../common/interfaces.sol";
+import { IUbiquityBondingV2, IUbiquityMetaPool, I3Pool } from "./interfaces.sol";
+import { Helpers } from "./helpers.sol";
+import { Events } from "./events.sol";
+
+contract ConnectV2Ubiquity is Helpers, Events {
```
Comment (thrilok209) [inline contracts/mainnet/connectors/ubiquity/main.sol]: @zapaz ConnectV2Ubiquity will have only name variable in it. All the public functions will go into `main`.

**Comment by zapaz** [comment]
Great @thrilok209  ! 

I will take into account your suggestions, and fix the checks failure...

File: contracts/mainnet/connectors/ubiquity/main.sol (hunk 1 above)
Comment (zapaz) [reply contracts/mainnet/connectors/ubiquity/main.sol]: ok

**Review by thrilok209 — APPROVED** [review]
@zapaz Connector looks good! You can proceed to deploy the connector. And share the connector address in the comment.

**Comment by 0x4007** [comment]
https://etherscan.io/address/0x8ec066d75d665616a94f2eccdbe49b54eaeefc78#code

**Comment by thrilok209** [comment]
Connector enabled: https://etherscan.io/tx/0xe0a569d11471d494c47117c36432fcb5048b2f39c423af99860dd17a9cebe45c

**Comment by thrilok209** [comment]
@zapaz @pavlovcik Could you resolve the conflicts?

**Comment by zapaz** [comment]
Made a merge of main inside ubiquity branch, and solve package.json issues...

and you merge before I can comment :-)

**Pull request merged**